File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
       if(!head || !head->next){
            return true;
       }
       struct ListNode *slow=head;
       struct ListNode *fast=head;
       while(fast->next && fast->next->next){
            slow=slow->next;
            fast=fast->next->next;
       }
       struct ListNode *prev=nullptr;
       struct ListNode *curr = slow->next;
       while(curr){
            struct ListNode *temp = curr->next;
            curr->next=prev;
            prev=curr;
            curr=temp;
       }
       struct ListNode *firstHalf = head;
       struct ListNode *secondHalf = prev;
       while(secondHalf){
            if(firstHalf->val!=secondHalf->val){
                return false;
            }
            firstHalf=firstHalf->next;
            secondHalf=secondHalf->next;
       }
       return true;
    }
};

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if(head==null || head.next== null) return true;
        ListNode slow = head;
        ListNode fast = head;
        while(fast!=null && fast.next!=null){
            fast = fast.next.next;
            slow = slow.next;
        }
        ListNode secondhalf = reverse(slow);

        ListNode p1=head,p2=secondhalf;
        while(p2!=null){
            if(p1.val!=p2.val){
                return false;
            }
            p1 = p1.next;
            p2 = p2.next;
        }
        return true;
    }
    public ListNode reverse(ListNode slow){
        ListNode prev = null;
        ListNode curr = slow;
        while(curr!=null){
            ListNode temp = curr;
            curr = curr.next;
            temp.next = prev;
            prev = temp;
        }
        return prev;
    }
}